app/services/linked_document_resolver.rb

```diff
@@ -0,0 +1,98 @@
+    # @param siblings [Array<String>] the identifiers for the related records
+    # @param solr_field [String] the Solr field containing the sibling values
+    def initialize(root:, siblings:, solr_field:)
+      @root = root
+      @siblings = siblings
+      @field = solr_field
+    end
+
+    # Retrieves the Solr documents for each adjacent node
+    # @return [Array<SolrDocument>] sibling documents
+    def siblings
+      return [] unless response.key?('response') && !response['response']['docs'].empty?
+
+      sibling_docs = response['response']['docs'].reject { |document| document['id'] == @root }
+      sibling_docs.map { |document| SolrDocument.new(document) }
+    end
+
+    # Decorate the SolrDocument for each sibling node decorated for the View layer
+    # @param title_field [String] field used to access the value of the title for the catalog record
+    # @param display_fields [Array<String>] the set of fields exposed for the catalog record
+    # @return [Array<SolrDocumentDecorator>] sibling document
+    def decorated(title_field: 'title_display', display_fields: [])
+      siblings.map do |sibling|
+        DecoratorService::SolrDocumentDecorator.new(document: sibling,
+                                                    title_field: title_field,
+                                                    display_fields: display_fields)
+      end
+    end
+
+    private
+
+      # Determine whether or not this number is a BIBID
+      # @param sibling_number [String] reference number for a linked catalog record
+      # @return [MatchData,nil] a match containing the ID for the linked BIBID (or nil)
+      def bib_id?(sibling_number)
+        /(?:^BIB)(.*)/.match(sibling_number)
+      end
+
+      # The Solr field containing the reference number for a linked catalog record
+      # @param sibling_number [String] reference number for a linked catalog record
+      # @return [String] the field
+      def solr_field(sibling_number)
+        bib_id?(sibling_number) ? 'id' : @field
+      end
+
+      # The value for the Solr field referencing the linked catalog record
+      # @param sibling_number [String] reference number for a linked catalog record
+      # @return [String] the value
+      def solr_value(sibling_number)
+        bib_match = bib_id?(sibling_number)
+        bib_match ? bib_match[1] : sibling_number
+      end
+
+      # The query for used to retrieve Documents for all catalog records linked to a given record
+      # @return [String] the query
+      def facet_query
+        queries = @siblings.map do |sibling_number|
+          "#{solr_field(sibling_number)}:#{solr_value(sibling_number)}"
+        end
+        queries.join(' OR ')
+      end
+
+      # Retrieve a response from the Solr endpoint for a faceted query
+      # @param fq [String] the Solr facet query
+      # @return [Faraday::Response] the HTTP response to the query
+      def get_fq_solr_response(fq)
```

Comment:
This should be deduplicated with https://github.com/pulibrary/orangelight/pull/1038/files#diff-65c43d8c8cb6c83aaf3fe8d18991633aR221